fork_choice/confirmation-rule.md

```diff
@@ -0,0 +1,673 @@
+    - [`get_remaining_weight_in_current_epoch`](#get_remaining_weight_in_current_epoch)
+    - [`get_leaf_block_roots`](#get_leaf_block_roots)
+    - [`get_current_epoch_participating_indices`](#get_current_epoch_participating_indices)
+    - [`get_ffg_support`](#get_ffg_support)
+    - [`is_ffg_confirmed_current_epoch`](#is_ffg_confirmed_current_epoch)
+    - [`is_ffg_confirmed_previous_epoch`](#is_ffg_confirmed_previous_epoch)
+    - [`is_ffg_confirmed`](#is_ffg_confirmed)
+  - [`is_confirmed`](#is_confirmed)
+- [Safe Block Hash](#safe-block-hash)
+  - [Helper Functions](#helper-functions-1)
+    - [`find_confirmed_block`](#find_confirmed_block)
+  - [`get_safe_execution_payload_hash`](#get_safe_execution_payload_hash)
+- [Confirmation Score](#confirmation-score)
+  - [Helper Functions](#helper-functions-2)
+    - [`get_one_confirmation_score`](#get_one_confirmation_score)
+    - [`get_lmd_confirmation_score`](#get_lmd_confirmation_score)
+    - [`get_ffg_confirmation_score_current_epoch`](#get_ffg_confirmation_score_current_epoch)
+    - [`get_ffg_confirmation_score_previous_epoch`](#get_ffg_confirmation_score_previous_epoch)
+    - [`get_ffg_confirmation_score`](#get_ffg_confirmation_score)
+  - [`get_confirmation_score`](#get_confirmation_score)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+<!-- /TOC -->
+
+## Introduction
+
+This document specifies a fast block confirmation rule for the Ethereum protocol.
+*Note:* Confirmation is not a substitute for finality! The safety of confirmations is weaker than that of finality.
+
+The research paper for this rule is attached in this [ethresear.ch post](https://ethresear.ch/t/confirmation-rule-for-ethereum-pos/15454).
+
+This rule makes the following network synchrony assumption: starting from the current slot, attestations created by honest validators in any slot are received by the end of that slot.
+Consequently, this rule provides confirmations to users who believe in the above assumption. If this assumption is broken, confirmed blocks can be reorged without any adversarial behavior, and without slashing.
+
+There are two algorithms in the document:
+- [**Confirmation Rule**](#confirmation-rule): Given a block and confirmation safety parameters, outputs whether the block is confirmed.
+- [**Confirmation Score**](#confirmation-score): Given a block, outputs the confirmation score for the block, i.e., the maximum possible confirmation safety parameters to deem the block confirmed.
+
+*Note:* These algorithms use unbounded integer arithmetic in some places. The rest of `consensus-specs` uses `uint64` arithmetic exclusively to ensure that results fit into length-limited fields - a property crucial for consensus objects (such as the `BeaconBlockBody`). This document describes a local confirmation rule that does not require storing anything in length-limited fields. Using unbounded integer arithmetic here prevents possible overflowing issues.
```

Comment:
For example, the multiplication or additions of two '`uint64` values should be stored into a `unit128` to make sure that no overflow can occur. Unless, of course, the two values have upperbounds lower than the maximum value that can be stored into a `uint64` and such that one is sure that the operation can never overflow.

The overarching reason for using unbounded ints here so that the spec can abstract away from such issues which, I think, are implementation concerns, not spec concerns.

Also, if we use unbounded ints in the specs, we know that, with respect to overflow, the generated reference tests are correct and therefore can detect overflow issues in the implementations. On the other hand, if we use bounded types in the spec and implementations use the same types, but we make a mistake and some of the calculations end up overflowing, then we have no way to catch such issues in the implementations (as the implementations copied the spec).

Does the above make sense?

Reply:
It does make sense to me to use unbounded `int` whenever we deal with total balance.

If total balance is 100M ETH then its Gwei representation takes 57 bits. A number of spare bits left with `uint64` is 7 which is quite risky in the perspective of overflow.